libcst/metadata/full_repo_manager.py

```diff
@@ -0,0 +1,154 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+# pyre-strict
+
+
+import json
+import subprocess
+from pathlib import Path
+from typing import (
+    TYPE_CHECKING,
+    Callable,
+    Collection,
+    Dict,
+    List,
+    Mapping,
+    Optional,
+    Sequence,
+    Tuple,
+)
+
+from mypy_extensions import TypedDict
+
+import libcst as cst
+from libcst.metadata.type_inference_provider import (
+    InferredType,
+    PyreData,
+    TypeInferenceProvider,
+)
+from libcst.metadata.wrapper import MetadataWrapper
+
+
+if TYPE_CHECKING:
+    from libcst.metadata.base_provider import ProviderT
+
+
+class FullRepoManager:
+    def __init__(
+        self,
+        repo_root_dir: str,
+        paths: Collection[str],
+        providers: Collection["ProviderT"],
+        timeout: int = 5,
+    ) -> None:
+        """
+        Given project root directory with pyre and watchman setup, :class:`~libcst.metadata.FullRepoManager`
+        handles the inter process communication to read the required full repository cache data for
+        metadata provider like :class:`~libcst.metadata.TypeInferenceProvider`.
+
+        :param paths: a collection of paths to access full repository data.
+        :param providers: a collection of metadata provider classes require accessing full repository
+            data, currently supports :class:`~libcst.metadata.TypeInferenceProvider`.
+        :param timeout: number of seconds. Raises `TimeoutExpired <https://docs.python.org/3/library/subprocess.html#subprocess.TimeoutExpired>`_
+            when timeout.
+        """
+        self.root_path: Path = Path(repo_root_dir)
+        self._cache: Dict["ProviderT", Mapping[str, object]] = {}
```

Comment:
I  guess there's no way of specifying that the type of the value of the map is whatever is provided by `ProviderT`?

Reply:
We didn't think of  an easy way to do that. One idea we had to annotate the type of cache for each provider is to add a second generic type variable in provider class definition for the cache type. This approach requires inspect the second generic type at runtime which make it complex.